crates/biome_js_analyze/src/lint/nursery/no_secrets.rs

```diff
@@ -0,0 +1,254 @@
+    (Pattern::Regex(&FACEBOOK_OAUTH_REGEX), "Facebook OAuth"),
+    (Pattern::Regex(&GOOGLE_OAUTH_REGEX), "Google OAuth"),
+    (Pattern::Regex(&AWS_API_KEY_REGEX), "AWS API Key"),
+    (Pattern::Regex(&HEROKU_API_KEY_REGEX), "Heroku API Key"),
+    (Pattern::Regex(&PASSWORD_IN_URL_REGEX), "Password in URL"),
+    (
+        Pattern::Regex(&GOOGLE_SERVICE_ACCOUNT_REGEX),
+        "Google (GCP) Service-account",
+    ),
+    (Pattern::Regex(&TWILIO_API_KEY_REGEX), "Twilio API Key"),
+    (
+        Pattern::Contains("-----BEGIN RSA PRIVATE KEY-----"),
+        "RSA Private Key",
+    ),
+    (
+        Pattern::Contains("-----BEGIN OPENSSH PRIVATE KEY-----"),
+        "SSH (OPENSSH) Private Key",
+    ),
+    (
+        Pattern::Contains("-----BEGIN DSA PRIVATE KEY-----"),
+        "SSH (DSA) Private Key",
+    ),
+    (
+        Pattern::Contains("-----BEGIN EC PRIVATE KEY-----"),
+        "SSH (EC) Private Key",
+    ),
+    (
+        Pattern::Contains("-----BEGIN PGP PRIVATE KEY BLOCK-----"),
+        "PGP Private Key Block",
+    ),
+];
+
+// TODO: Try to get this to work in JavaScript comments as well
+declare_lint_rule! {
+    /// Disallow usage of sensitive data such as API keys and tokens.
+    ///
+    /// This rule checks for high-entropy strings and matches common patterns
+    /// for secrets, such as AWS keys, Slack tokens, and private keys.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// var secret = "AKIA1234567890EXAMPLE";
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// var nonSecret = "hello world";
+    /// ```
+    pub NoSecrets {
+        version: "next",
+        name: "noSecrets",
+        language: "js",
+        recommended: false,
+        sources: &[RuleSource::Eslint("no-secrets/no-secrets")],
+        source_kind: RuleSourceKind::Inspired,
+    }
+}
+
+impl Rule for NoSecrets {
+    type Query = Ast<JsStringLiteralExpression>;
+    type State = &'static str;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let token = node.value_token().ok()?;
+        let text = Arc::new(token.text().to_string());
+
+        let result = Arc::new(Mutex::new(None));
+        let found = Arc::new(Mutex::new(false));
+        let num_threads = 4;
+        let patterns_per_thread = (SENSITIVE_PATTERNS.len() + num_threads - 1) / num_threads;
+
+        // Since regex matching is expensive, we run in threads
+        // - Added a mutex for "found", i.e. I want to exit early if any
+        // pattern matches or string contains is found
+        // - Currently hardcoded to 4 threads. Can be adjusted later.
+        // - I do realize that it might be overkill for smaller strings.
+        // Maybe can check that in future.
+        let handles: Vec<_> = SENSITIVE_PATTERNS
+            .chunks(patterns_per_thread)
+            .map(|chunk| {
```

Comment:
This seems like a premature optimization. Biome already processes files in parallel. I would rather go with the naive implementation unless you have benchmarked this and it shows a significant improvement.

For reference though, you should use the `rayon` crate for this sort of thing because it handles a lot of boilerplate.

Reply:
Noob question: How can I manually retrigger the benchmark we have in CI?

Reply:
I need help converting this to `rayon`.

I can't seem to get it to work in `no_secrets.rs`:
```zsh
$ just test-lintrule noSecrets

just _touch crates/biome_js_analyze/tests/spec_tests.rs
touch crates/biome_js_analyze/tests/spec_tests.rs
just _touch crates/biome_json_analyze/tests/spec_tests.rs
touch crates/biome_json_analyze/tests/spec_tests.rs
just _touch crates/biome_css_analyze/tests/spec_tests.rs
touch crates/biome_css_analyze/tests/spec_tests.rs
just _touch crates/biome_graphql_analyze/tests/spec_tests.rs
touch crates/biome_graphql_analyze/tests/spec_tests.rs
cargo test -p biome_js_analyze -- no_secrets --show-output
   Compiling biome_js_analyze v0.5.7 (/Users/saadbazaz/Projects/biome/crates/biome_js_analyze)
error[E0433]: failed to resolve: use of undeclared crate or module `rayon`
  --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:12:5
   |
12 | use rayon::prelude::*;
   |     ^^^^^ use of undeclared crate or module `rayon`

error[E0599]: no method named `par_iter` found for reference `&'static [(no_secrets::Pattern, &'static str)]` in the current scope
  --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:67:28
   |
67 |         SENSITIVE_PATTERNS.par_iter().find_any(|&&(pattern, comment)| {
   |                            ^^^^^^^^
   |
help: there is a method `iter` with a similar name
   |
67 |         SENSITIVE_PATTERNS.iter().find_any(|&&(pattern, comment)| {
   |                            ~~~~

Some errors have detailed explanations: E0433, E0599.
For more information about an error, try `rustc --explain E0433`.
error: could not compile `biome_js_analyze` (lib) due to 2 previous errors
error: Recipe `test-lintrule` failed on line 128 with exit code 101
```

My code:
```rs
use biome_rowan::AstNode;
use regex::Regex;

use std::sync::{Arc, LazyLock, Mutex};
use rayon::prelude::*; // <-- added this

enum Pattern {
    Regex(&'static LazyLock<Regex>),
    Contains(&'static str),
}

// ...

fn run(ctx: &RuleContext<Self>) -> Self::Signals {
        let node = ctx.query();
        let token = node.value_token().ok()?;
        let text = token.text().to_string();

        if text.len() < 24 {
            return None;
        }

        let result = Arc::new(Mutex::new(None));
        let found = Arc::new(Mutex::new(false));

        SENSITIVE_PATTERNS.par_iter().find_any(|&&(pattern, comment)| {
            if *found.lock().unwrap() {
                return false;
            }
            let matched = match pattern {
                Pattern::Regex(re) => re.is_match(&text),
                Pattern::Contains(substring) => text.contains(substring),
            };
            if matched {
                let mut guard = result.lock().unwrap();
                if guard.is_none() {
                    *guard = Some(comment);
                    *found.lock().unwrap() = true;
                }
                true
            } else {
                false
            }
        });

        let guard = result.lock().unwrap();
        if let Some(comment) = *guard {
            return Some(comment);
        }

        if is_high_entropy(&text) {
            Some("The string has a high entropy value")
        } else {
            None
        }
    }
    
    // ...
```

Reply:
The error here is:

> failed to resolve: use of undeclared crate or module `rayon`

`biome_js_analyze` doesn't have `rayon` in it's `Cargo.toml`

Reply:
>  How can I manually retrigger the benchmark we have in CI?

Take a look at `xtask/bench/README.md`

Reply:
> The error here is:
> 
> > failed to resolve: use of undeclared crate or module `rayon`
> 
> `biome_js_analyze` doesn't have `rayon` in it's `Cargo.toml`

Oh... I was looking at another package.

Do you think I should add it? It'd be a first...

Reply:
Yeah, you can go ahead and add `regex = { workspace = true }` to the dependencies, and use `just format` to format everything, including the toml files.

Reply:
Oof, purged it for now. Based on [this comment](https://github.com/biomejs/biome/pull/3823#pullrequestreview-2290136899). Let's test without it, you're right it's a premature optimization.